processing/src/main/java/io/druid/query/groupby/epinephelinae/EpiGroupByStrategy.java

```diff
@@ -0,0 +1,205 @@
+    this.mergeBufferPool = mergeBufferPool;
+    this.spillMapper = spillMapper;
+    this.queryWatcher = queryWatcher;
+  }
+
+  @Override
+  public Sequence<Row> mergeResults(
+      final QueryRunner<Row> baseRunner,
+      final GroupByQuery query,
+      final Map<String, Object> responseContext
+  )
+  {
+    // Merge streams using ResultMergeQueryRunner, then apply postaggregators, then apply limit (which may
+    // involve materialization)
+
+    final ResultMergeQueryRunner<Row> mergingQueryRunner = new ResultMergeQueryRunner<Row>(baseRunner)
+    {
+      @Override
+      protected Ordering<Row> makeOrdering(Query<Row> queryParam)
+      {
+        return ((GroupByQuery) queryParam).getRowOrdering(true);
+      }
+
+      @Override
+      protected BinaryFn<Row, Row, Row> createMergeFn(Query<Row> queryParam)
+      {
+        return new EpiGroupByBinaryFn((GroupByQuery) queryParam);
+      }
+    };
+
+    // Fudge timestamp, maybe. Necessary to keep timestamps in sync across partial queries.
+    final QueryGranularity gran = query.getGranularity();
+    final String fudgeTimestamp;
+    if (query.getContextValue(CTX_KEY_FUDGE_TIMESTAMP, "").isEmpty() && gran instanceof AllGranularity) {
+      final long timeStart = query.getIntervals().get(0).getStartMillis();
+      fudgeTimestamp = String.valueOf(
+          new DateTime(gran.iterable(timeStart, timeStart + 1).iterator().next()).getMillis()
+      );
+    } else {
+      fudgeTimestamp = query.getContextValue(CTX_KEY_FUDGE_TIMESTAMP, "");
+    }
+
+    return query.applyLimit(
+        Sequences.map(
+            mergingQueryRunner.run(
+                new GroupByQuery(
+                    query.getDataSource(),
+                    query.getQuerySegmentSpec(),
+                    query.getDimFilter(),
+                    gran,
+                    query.getDimensions(),
+                    query.getAggregatorSpecs(),
+                    // Don't do post aggs until the end of this method.
+                    ImmutableList.<PostAggregator>of(),
+                    // Don't do "having" clause until the end of this method.
+                    null,
+                    null,
+                    query.getContext()
+                ).withOverriddenContext(
+                    ImmutableMap.<String, Object>of(
+                        "finalize", false,
+                        GroupByStrategySelector.CTX_KEY_STRATEGY, GroupByStrategySelector.STRATEGY_EPINEPHELINAE,
+                        CTX_KEY_FUDGE_TIMESTAMP, fudgeTimestamp
+                    )
+                ),
+                responseContext
+            ),
+            new Function<Row, Row>()
+            {
+              @Override
+              public Row apply(final Row row)
+              {
+                // Maybe apply postAggregators.
+
+                if (query.getPostAggregatorSpecs().isEmpty()) {
+                  return row;
+                }
+
+                final Map<String, Object> newMap;
```

Comment:
It's not too bad if you don't have postAggregators. If you do then you have to copy the map and that adds a bit of time, although not really any more than the old strategy would incur in the same situation. I guess we could optimize this in the future if it ends up being an issue.